platform/security/core/security-core-impl/src/main/java/ddf/security/samlp/impl/EntityInformation.java

```diff
@@ -0,0 +1,270 @@
+import org.opensaml.saml2.core.AuthnRequest;
+import org.opensaml.saml2.metadata.AssertionConsumerService;
+import org.opensaml.saml2.metadata.Endpoint;
+import org.opensaml.saml2.metadata.EntityDescriptor;
+import org.opensaml.saml2.metadata.KeyDescriptor;
+import org.opensaml.saml2.metadata.SPSSODescriptor;
+import org.opensaml.xml.schema.XSBase64Binary;
+import org.opensaml.xml.security.credential.UsageType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableSet;
+
+import ddf.security.samlp.SamlProtocol;
+import ddf.security.samlp.SamlProtocol.Binding;
+
+@Immutable
+public class EntityInformation {
+    private static final Logger LOGGER = LoggerFactory.getLogger(EntityInformation.class);
+
+    private static final ImmutableSet<Binding> SUPPORTED_BINDINGS =
+            ImmutableSet.of(Binding.HTTP_POST, Binding.HTTP_REDIRECT);
+
+    private static final Binding PREFERRED_BINDING = Binding.HTTP_REDIRECT;
+
+    private final String signingCertificate;
+
+    private final String encryptionCertificate;
+
+    private final ServiceInfo defaultAssertionConsumerService;
+
+    private final Map<Binding, ServiceInfo> assertionConsumerServices;
+
+    private final Map<Binding, ServiceInfo> logoutServices;
+
+    private EntityInformation(Builder builder) {
+        signingCertificate = builder.signingCertificate;
+        encryptionCertificate = builder.encryptionCertificate;
+        defaultAssertionConsumerService = builder.defaultAssertionConsumerService;
+        assertionConsumerServices = builder.assertionConsumerServices;
+        logoutServices = builder.logoutServices;
+    }
+
+    public String getSigningCertificate() {
+        return signingCertificate;
+    }
+
+    public String getEncryptionCertificate() {
+        return encryptionCertificate;
+    }
+
+    public ServiceInfo getLogoutService() {
+        return getLogoutService(null);
+    }
+
+    public ServiceInfo getLogoutService(Binding preferred) {
+        Binding binding = getBinding(null, preferred);
+        ServiceInfo logoutServiceInfo = logoutServices.get(binding);
+        if (logoutServiceInfo == null) {
+            logoutServiceInfo = logoutServices.values()
+                    .stream()
+                    .findFirst()
+                    .orElse(null);
+        }
+
+        return logoutServiceInfo;
+    }
+
+    Binding getBinding(AuthnRequest request, Binding preferred) {
+        if (request != null && request.getProtocolBinding() != null && SUPPORTED_BINDINGS.contains(
+                Binding.from(request.getProtocolBinding()))) {
+            return Binding.from(request.getProtocolBinding());
+        }
+
+        return preferred != null ? preferred : PREFERRED_BINDING;
+    }
+
+    public ServiceInfo getAssertionConsumerService(AuthnRequest request, Binding preferred) {
+        ServiceInfo si = null;
+        Binding binding = preferred != null ? preferred : PREFERRED_BINDING;
```

Comment:
This line can be moved below the following `if` block. Then this code becomes very similar to the `getBinding` method above. It might be possible to refactor at that point.